### ISSUE TITLE
Update Vim from 7.4.52 to 7.4.399+

### ISSUE THREAD

**Issue by nancym** [Microsoft/BashOnWindows#543]
I use Vim 7.4.399 or higher on all my devices so I can encrypt and decrypt files using the blowfish2 cryptmethod. I do this by putting the following in my `vimrc`:

```
set cryptmethod=blowfish2
```

The default version of Vim on BashOnWindows is 7.4.52 and it does not support blowfish2. It does support either of these:

```
set cryptmethod=zip
set cryptmethod=blowfish
```

I'd like to be able to edit all my blowfish2 encrypted files from CLI Vim within BashOnWindows (instead of having to use gVim (GUI Vim)). I have two questions:
- How do I update my BashOnWindows version of Vim?
- What should I do to encourage Canonical to include Vim 7.4.399 or higher as the default version of Vim in BashOnWindows?

Thank you!

**Comment by zerocool4u2** [comment]
i guest you could directly download it from their github and build it yourself, http://www.vim.org/git.php, for what i was reading they are at 7.4.1941, or you can just add another untrusted ppa https://launchpad.net/ubuntu/+ppas?name_filter=vim, it usually takes some time to canonical to get their packages current on stable versions, unless there are mayor changes (like vim 8 soon), also the current lte of ubuntu is 16.04, maybe they get a newer version of vim, in your case i would add some ppa with more frequent updates of the build.

**Comment by aseering** [comment]
Canonical doesn't generally put major package updates into an existing release.  Instead, it includes them in the next release.

Microsoft(?) is using an older version of Ubuntu, 14.04.  (It was less old when this project started :-) )  The latest version is 16.04, which contains vim 7.4.1689.  So the most-likely way to get your update bundled is for Bash On Windows to upgrade to Ubuntu 16.04.  This has been discussed elsewhere in this bug tracker; it sounds like it is probably on their ToDo list but to my knowledge no timeframe has been given.

In the meantime, @zerocool4u2 's suggestions are probably the best workaround.

**Comment by therealkenc** [comment]
```
sudo add-apt-repository ppa:pkg-vim/vim-daily
sudo apt-get update && sudo apt-get upgrade
```
Finding the magic PPA for the app in question is the standard way to deal with this general problem for people stuck on 14.04 or 12.04 (often because they're stuck in the cloud). 

I wonder if MS could get together with Canonical and arrange for a variant of `do-release-upgrade` that just leaves the kernel alone. Sure, Linus releases new kernels all the time, but for the purposes of WSL, the Linux syscall interface has been stable for years. [For WSL's stated purpose of running sed, awk, etc it's been 32-bit stable for decades.]

Maybe there's some dpk-fu to get the Xenial binaries for userland tools. I did some googling around for an answer to that, but for anyone running native Ubuntu the question isn't even well formed.

**Comment by aseering** [comment]
#551 was filed by someone who sounds like they're trying to apply the [Debian dist-upgrade instructions](https://www.debian.org/releases/stable/i386/release-notes/ch-upgrading.en.html) to Ubuntu, which should in theory mostly achieve what you're trying to do.  Though you would probably have to pin the package containing the kernel at its current version and you would probably want to manually apply any packaging fixes that are hardcoded into the `do-release-upgrade` script.

Note that WSL only runs 64-bit binaries, even sed/awk/etc.  Also, [WSL's stated purpose] also includes running NodeJS, Python, Perl, and Git, among other things.  While it is true that the syscall interface is pretty stable, the filesystem interfaces to the kernel (/proc and /sys) have been changing much more rapidly.  (And keep in mind that WSL's implementation of the Linux syscall interface is incomplete; it's possible that a new version of some program might use a syscall with different and unsupported arguments as compared to the old version.)  #540 tracks one example of an updated tool that tries to use a filesystem-based kernel interface that's not yet present in WSL.

**Comment by PaulSchulz** [comment]
After upgrading using 'apt-get' to upgrade to 16.04 vim is at 7.4.1689. (apt-cache show vim)

**Comment by PaulSchulz** [comment]
There is no kernel package to pin.

**Comment by therealkenc** [comment]
@aseering Yes, WSL only runs 64-bit binaries; which is why I needed to state the 32-bit caveat to say "decades". A new version of NodeJS might trip over some unimplemented WSL surface, but nothing in NodeJS or any WSL scoped software requires any **new** kernel surface area that has changed since Ubuntu 14.04 was released, including `/dev` or `/proc`. I am not saying `/proc` or `/dev` are forever immutable, because obviously they're not; but for presently WSL-scoped software, they are not changing appreciably at all. It is worth noting per #540 that `/proc/sys/vm/min_free_kbytes` dates back to _at least_ the [April 2005](https://github.com/torvalds/linux/blob/1da177e4c3f41524e886b7f1b8a0c1fc7321cac2/kernel/sysctl.c) 2.6.12-rc2 kernel (as far as git goes back).
